samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -64,43 +61,51 @@
   private static final String APPLICATION_RUNNER_PATH_SUFFIX = "/ApplicationRunnerData";
 
   private final String uid;
+  private final LocalJobPlanner planner;
   private final Set<StreamProcessor> processors = ConcurrentHashMap.newKeySet();
   private final CountDownLatch shutdownLatch = new CountDownLatch(1);
   private final AtomicInteger numProcessorsToStart = new AtomicInteger();
   private final AtomicReference<Throwable> failure = new AtomicReference<>();
-  private final Map<String, MetricsReporter> customMetricsReporters;
 
   private ApplicationStatus appStatus = ApplicationStatus.New;
 
-  final class LocalStreamProcessorLifeCycleListener implements StreamProcessorLifecycleListener {
-    StreamProcessor processor;
+  /**
+   * Defines a specific implementation of {@link ProcessorLifecycleListener} for local {@link StreamProcessor}s.
+   */
+  private final class LocalStreamProcessorLifecycleListener implements ProcessorLifecycleListener {
+    private final StreamProcessor processor;
+    private final ProcessorLifecycleListener processorLifecycleListener;
 
-    void setProcessor(StreamProcessor processor) {
-      this.processor = processor;
+    @Override
+    public void beforeStart() {
+      processorLifecycleListener.beforeStart();
     }
 
     @Override
-    public void onStart() {
+    public void afterStart() {
+      processorLifecycleListener.afterStart();
```

Comment:
Make sense. Actually, changing the invocation of all processorLifecycleListener methods to be after setting the ApplicationStatus in shutdownAndNotify(), before shutdownLatch.countDown(). That ways, it will guarantee that the user's listener callbacks are invoked before the external caller for ApplicationRunner.waitForFinish() returns. Hence, any customized final status user code want to set in the callbacks are guaranteed to be available after runner.waitForFinish().